internal/handler/job/events/events.go

```diff
@@ -0,0 +1,138 @@
+package events
+
+/*
+ * SPDX-FileCopyrightText: 2023 Siemens AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Author: Michael Adler <michael.adler@siemens.com>
+ */
+
+import (
+	"context"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/google/uuid"
+	"github.com/olebedev/emitter"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"github.com/siemens/wfx/generated/model"
+	"github.com/siemens/wfx/middleware/logging"
+)
+
+type JobEvent struct {
+	// Ctime is the time when the event was created
+	Ctime  strfmt.DateTime `json:"ctime"`
+	Action Action          `json:"action"`
+	Job    *model.Job      `json:"job"`
+}
+
+type FilterParams struct {
+	JobIDs    []string
+	ClientIDs []string
+	Workflows []string
+}
+
+type Action string
+
+const (
+	ActionCreate           Action = "CREATE"
+	ActionDelete           Action = "DELETE"
+	ActionAddTags          Action = "ADD_TAGS"
+	ActionDeleteTags       Action = "DELETE_TAGS"
+	ActionUpdateStatus     Action = "UPDATE_STATUS"
+	ActionUpdateDefinition Action = "UPDATE_DEFINITION"
+)
+
+var e *emitter.Emitter = emitter.New(32)
+
+// AddSubscriber adds a new subscriber to receive job events filtered based on the provided filterParams.
+func AddSubscriber(ctx context.Context, filter FilterParams) (<-chan emitter.Event, error) {
+	log := logging.LoggerFromCtx(ctx)
+
+	// for logging purposes
+	subscriberID := uuid.New().String()
+
+	log.Info().
+		Str("subscriberID", subscriberID).
+		Dict("filterParams", zerolog.Dict().
+			Strs("clientIDs", filter.ClientIDs).
+			Strs("jobIDs", filter.JobIDs).
+			Strs("workflows", filter.Workflows)).
+		Msg("Adding new subscriber for job events")
+
+	if mw := filter.toMiddleware(subscriberID); mw != nil {
+		return e.On("*", mw), nil
+	}
+	return e.On("*"), nil
+}
+
+// ShutdownSubscribers disconnects all subscribers.
+func ShutdownSubscribers() {
+	// note: e.Off("*") does not match topics containing the character '/'
+	// since the library uses a glob and tokens are separated by '/'
+	for _, topic := range e.Topics() {
+		log.Debug().Str("topic", topic).Msg("Closing subscribers")
+		e.Off(topic)
+	}
+	log.Info().Msg("Unsubscribed all subscribers")
+}
+
+// SubscriberCount counts the total number of subscribers across all topics.
+func SubscriberCount() int {
+	count := 0
+	for _, topic := range e.Topics() {
+		count += len(e.Listeners(topic))
+	}
+	return count
+}
+
+// PublishEvent publishes a new event in a *synchronous* manner.
+func PublishEventSync(event JobEvent) {
+	<-e.Emit(event.Job.ID, event)
+}
+
+// PublishEvent publishes a new event in an *asynchronous* manner.
+func PublishEvent(event JobEvent) {
+	go func() {
+		PublishEventSync(event)
+	}()
+}
+
+func (filter FilterParams) toMiddleware(subscriberID string) func(*emitter.Event) {
```

Comment:
Hm, isn't there a better name as `toMiddleware()` for this functionality?

Reply:
Renamed to `createEventFilter`